blackjax/tempered_smc.py

```diff
@@ -0,0 +1,251 @@
+    fun: Callable
+        The decreasing function to solve, we must have fun(min_delta) > 0, fun(max_delta) < 0
+    min_delta: float
+        Starting point of the interval search
+    max_delta: float
+        End point of the interval search
+    eps: float
+        Tolerance for |f(x)|
+
+    Returns
+    -------
+    delta: jnp.ndarray, shape (,)
+        The root of `fun`
+
+    """
+
+    def body(carry):
+        a, b, f_a, f_b = carry
+        mid = 0.5 * (a + b)
+        f_mid = fun(mid)
+        a, b, f_a, f_b = jax.lax.cond(f_mid < 0,
+                                      lambda _: (a, mid, f_a, f_mid),
+                                      lambda _: (mid, b, f_mid, f_b),
+                                      None)
+        return a, b, f_a, f_b
+
+    def cond(carry):
+        a, b, f_a, f_b = carry
+        return f_a - f_b > eps
+
+    f_min_delta, f_max_delta = fun(min_delta), fun(max_delta)
+
+    if_no_opt = lambda _: max_delta
+
+    def if_opt(_):
+        res_a, res_b, fun_res_a, fun_res_b = jax.lax.while_loop(cond, body,
+                                                                (min_delta, max_delta, f_min_delta, f_max_delta))
+        return 0.5 * (res_a + res_b)
+
+    # if the upper end of the interval returns positive already, just return it, otherwise search the optimum.
+    return jax.lax.cond(f_max_delta > 0, if_no_opt, if_opt, None)
+
+
+def ess_solver(potential_fn: Callable, smc_state: SMCState, target_ess: float,
+               root_solver: Callable = dichotomy_solver):
+    """Build a Tempered SMC step.
+
+    Parameters
+    ----------
+    potential_fn: Callable
+        A function that returns the potential energy of a chain at a given position.
+    smc_state: SMCState
+        Current state of the tempered SMC algorithm
+    target_ess: float
+        The relative ESS targeted for the next increment of SMC tempering
+    root_solver: Callable, optional
+        A solver to find the root of a function, takes a function `f`, a starting point `delta0`,
+        a min value `min_delta`, and a max value `max_delta`.
+        Default is `BFGS` minimization of `f ** 2` and ignores `min_delta` and `max_delta`.
+
+    Returns
+    -------
+    delta: float
+        The increment that
+    """
+
+    particles = smc_state.particles
+    n_particles = jax.tree_flatten(particles)[0][0].shape[0]
+
+    potential_val = potential_fn(smc_state.particles)
+
+    def fun_to_solve(delta):
+        log_ess_val = _log_ess(-delta * potential_val)
+        return log_ess_val - np.log(n_particles * target_ess)
+
+    estimated_delta = root_solver(fun_to_solve, 0., 0., 1 - smc_state.lamda)
+    next_lamda = smc_state.lamda + estimated_delta
+
+    delta = jax.lax.cond(next_lamda > 1,
+                         lambda _: 1 - smc_state.lamda,
+                         lambda _: estimated_delta,
+                         operand=None)
+    return delta
+
+
+def systematic(weights: jnp.ndarray, key: jnp.ndarray):
```

Comment:
Could go in `inference/smc/resampling.py` since there will likely be several resampling schemes in the future. I would move everything hmc related to `inference/hmc`.